File: rapida-java/src/main/java/ai/rapida/sdk/RapidaClientOption.java
/*
 *  Copyright (c) 2024. Rapida
 *
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 *
 *  Author: Prashant <prashant@rapida.ai>
 *
 */

package ai.rapida.sdk;
import ai.rapida.sdk.enums.RapidaEnvironment;
import ai.rapida.sdk.enums.RapidaRegion;

import java.util.Optional;
import java.util.logging.Logger;


/**
 * @author prashant.srivastav
 */
public class RapidaClientOption {
    private String rapidaApiKey;
    private String rapidaEndpointUrl;

    private String rapidaAssistantUrl;
    private RapidaRegion rapidaRegion;
    private RapidaEnvironment rapidaEnvironment;
    private boolean isSecure;

    private static final String ENDPOINT_URL = "connect.rapida.ai";
    private static final String ASSISTANT_URL = "assistant.rapida.ai:8080";
    private static final Logger logger = Logger.getLogger(RapidaClientOption.class.getName());


    public RapidaClientOption(
            String apiKey,
            String endpointUrl,
            String rapidaAssistantUrl,
            RapidaEnvironment environment,
            RapidaRegion region,
            boolean isSecure) {
        this.rapidaApiKey = apiKey;
        this.rapidaEndpointUrl = endpointUrl;
        this.rapidaAssistantUrl= rapidaAssistantUrl;
        this.rapidaEnvironment = environment;
        this.rapidaRegion = region;
        this.isSecure = isSecure;
    }

    public RapidaClientOption() {
        this.rapidaApiKey = System.getenv("RAPIDA_API_KEY");
        this.rapidaEndpointUrl = ENDPOINT_URL;
        this.rapidaAssistantUrl = ASSISTANT_URL;
        this.rapidaEnvironment = RapidaEnvironment.PRODUCTION;
        this.rapidaRegion = RapidaRegion.ALL;
        this.isSecure = true;
    }

    public Optional<String> getRapidaApiKey() {
        return Optional.ofNullable(rapidaApiKey);
    }

    public Optional<String> getRapidaEndpointUrl() {
        // Return the rapidaEndpointUrl wrapped in an Optional if it's not null
        if (this.rapidaEndpointUrl != null) {
            return Optional.of(this.rapidaEndpointUrl);
        }

        // If rapidaEndpointUrl is null, get the environment variable, or return the default endpoint URL
        String endpointUrl = Optional.ofNullable(System.getenv("RAPIDA_ENDPOINT_URL")).orElse(ENDPOINT_URL);
        return Optional.of(endpointUrl);
    }

    public Optional<String> getAssistantUrl() {
        // Return the rapidaEndpointUrl wrapped in an Optional if it's not null
        if (this.rapidaEndpointUrl != null) {
            return Optional.of(this.rapidaEndpointUrl);
        }

        // If rapidaEndpointUrl is null, get the environment variable, or return the default endpoint URL
        String endpointUrl = Optional.ofNullable(System.getenv("RAPIDA_ENDPOINT_URL")).orElse(ENDPOINT_URL);
        return Optional.of(endpointUrl);
    }

    public Optional<RapidaRegion> getRapidaRegion() {
        if (rapidaRegion != null) return Optional.of(this.rapidaRegion);
        return Optional.of(RapidaRegion.ALL);
    }

    public Optional<RapidaEnvironment> getRapidaEnvironment() {
        if (this.rapidaEnvironment != null) return Optional.of(this.rapidaEnvironment);
        return Optional.of(RapidaEnvironment.PRODUCTION);
    }

    public boolean isSecure() {
        return isSecure;
    }

    public void setRapidaEndpointUrl(String rapidaEndpointUrl) {
        this.rapidaEndpointUrl = rapidaEndpointUrl;
    }

    public void setRapidaAssistantUrl(String rapidaAssistantUrl) {
        this.rapidaAssistantUrl = rapidaAssistantUrl;
    }

    public void setRapidaRegion(RapidaRegion rapidaRegion) {
        this.rapidaRegion = rapidaRegion;
    }

    public void setRapidaEnvironment(RapidaEnvironment rapidaEnvironment) {
        this.rapidaEnvironment = rapidaEnvironment;
    }

    public void setSecure(boolean secure) {
        isSecure = secure;
    }
}

File: rapida-golang/rapida/rapida_client_option.go
/*
 *  Copyright (c) 2024. Rapida
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in
 *  all copies or substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *  THE SOFTWARE.
 *
 *  Author: Prashant Srivastav
 *
 */

package rapida

import (
	"os"

	rapida_constants "github.com/rapidaai/rapida-sdk/rapida/constants"
)

type RapidaClientOption struct {
	RapidaApiKey       *string
	RapidaEndpointUrl  *string
	RapidaAssistantUrl *string
	RapidaRegion       *rapida_constants.RapidaRegion
	RapidaEnvironment  *rapida_constants.RapidaEnvironment
	IsSecure           bool
}

const (
	ENDPOINT_URL  = "connect.rapida.ai"
	ASSISTANT_URL = "assistant.rapida.ai:8080"
)

// NewRapidaClientOptionWithParams initializes RapidaClientOption with provided parameters.
func NewRapidaClientOptionWithParams(apiKey, endpointUrl, assistantUrl string, environment rapida_constants.RapidaEnvironment, region rapida_constants.RapidaRegion, isSecure bool) *RapidaClientOption {
	return &RapidaClientOption{
		RapidaApiKey:       &apiKey,
		RapidaEndpointUrl:  &endpointUrl,
		RapidaAssistantUrl: &assistantUrl,
		RapidaEnvironment:  &environment,
		RapidaRegion:       &region,
		IsSecure:           isSecure,
	}
}

// NewRapidaClientOption initializes RapidaClientOption with default parameters.
func NewRapidaClientOption() *RapidaClientOption {
	apiKey := os.Getenv("RAPIDA_API_KEY")
	return &RapidaClientOption{
		RapidaApiKey:       &apiKey,
		RapidaEndpointUrl:  strPtr(ENDPOINT_URL),
		RapidaAssistantUrl: strPtr(ASSISTANT_URL),
		RapidaEnvironment:  envPtr(rapida_constants.PRODUCTION),
		RapidaRegion:       regionPtr(rapida_constants.ALL),
		IsSecure:           true,
	}
}

// GetRapidaApiKey returns the API key.
func (o *RapidaClientOption) GetRapidaApiKey() *string {
	return o.RapidaApiKey
}

// GetRapidaEndpointUrl returns the endpoint URL.
func (o *RapidaClientOption) GetRapidaEndpointUrl() *string {
	if o.RapidaEndpointUrl != nil {
		return o.RapidaEndpointUrl
	}
	endpointUrl := os.Getenv("RAPIDA_ENDPOINT_URL")
	if endpointUrl == "" {
		endpointUrl = ENDPOINT_URL
	}
	return &endpointUrl
}

// GetAssistantUrl returns the assistant URL.
func (o *RapidaClientOption) GetAssistantUrl() *string {
	if o.RapidaAssistantUrl != nil {
		return o.RapidaAssistantUrl
	}
	assistantUrl := os.Getenv("RAPIDA_ASSISTANT_URL")
	if assistantUrl == "" {
		assistantUrl = ASSISTANT_URL
	}
	return &assistantUrl
}

// GetRapidaRegion returns the RapidaRegion.
func (o *RapidaClientOption) GetRapidaRegion() *rapida_constants.RapidaRegion {
	if o.RapidaRegion != nil {
		return o.RapidaRegion
	}
	return regionPtr(rapida_constants.ALL)
}

// GetRapidaEnvironment returns the RapidaEnvironment.
func (o *RapidaClientOption) GetRapidaEnvironment() *rapida_constants.RapidaEnvironment {
	if o.RapidaEnvironment != nil {
		return o.RapidaEnvironment
	}
	return envPtr(rapida_constants.PRODUCTION)
}

// SetRapidaEndpointUrl sets the endpoint URL.
func (o *RapidaClientOption) SetRapidaEndpointUrl(endpointUrl string) {
	o.RapidaEndpointUrl = &endpointUrl
}

// SetRapidaAssistantUrl sets the assistant URL.
func (o *RapidaClientOption) SetRapidaAssistantUrl(assistantUrl string) {
	o.RapidaAssistantUrl = &assistantUrl
}

// SetRapidaRegion sets the RapidaRegion.
func (o *RapidaClientOption) SetRapidaRegion(region rapida_constants.RapidaRegion) {
	o.RapidaRegion = &region
}

// SetRapidaEnvironment sets the RapidaEnvironment.
func (o *RapidaClientOption) SetRapidaEnvironment(environment rapida_constants.RapidaEnvironment) {
	o.RapidaEnvironment = &environment
}

// SetSecure sets the security flag.
func (o *RapidaClientOption) SetSecure(secure bool) {
	o.IsSecure = secure
}

// Utility functions to create pointers from values

func strPtr(s string) *string {
	return &s
}

func regionPtr(r rapida_constants.RapidaRegion) *rapida_constants.RapidaRegion {
	return &r
}

func envPtr(e rapida_constants.RapidaEnvironment) *rapida_constants.RapidaEnvironment {
	return &e
}
